### ISSUE TITLE
Multi-Image Input for Finetuning

### ISSUE THREAD

**Issue by monroye** [ByteDance-Seed/Bagel#145]
Hi, I want to congratulate you on your hard work!
I saw this[ issue](https://github.com/ByteDance-Seed/Bagel/issues/83) where you explain how to take two images as input, like Omnigen.

I'm trying to fine-tune Bagel specifically for this kind of task, and so far, I have created a custom dataloader that takes two input images and one input text to generate a target image.

What I'd like to know is if this is the correct way to fine-tune the model for this type of task, or if I'm missing something or misunderstanding Bagel's architecture. I would greatly appreciate it if you could share the code you used to train the model with multiple input images.

```
class EditingIterableDataset(
    InterleavedBaseIterableDataset, ParquetStandardIterableDataset
):
    """
    Custom dataset for editing that ensures a deterministic sequence:
    Template → Reference → Instruction → Target

    This guarantees the model always sees both input images before generating the target.
    """

    def parse_row(self, row):
        """
        Deterministic parsing for food editing:
        - Always processes Image1 first (conditioning)
        - Always processes Image2 second (conditioning)
        - Always processes instruction third (conditioning)
        - Always generates target image last (with loss)
        """
        # Initialize data structure
        data = self._init_data()

        # Add Image 1
        template_img = pil_img2rgb(Image.open(io.BytesIO(row["image_list"][0])))
        data = self._add_image(
            data,
            template_img,
            need_loss=False,  # No loss
            need_vae=True, 
            need_vit=True, 
            enable_cfg=True,  
        )

        # Add Image 2
        reference_img = pil_img2rgb(Image.open(io.BytesIO(row["image_list"][1])))
        data = self._add_image(
            data,
            reference_img,
            need_loss=False,  # No loss - this is conditioning input
            need_vae=True,  
            need_vit=True,  
            enable_cfg=True,  
        )

        # Add instruction
        instruction = row["instruction_list"][0]  # Direct access - simple list format
        
        data = self._add_text(
            data,
            instruction,
            need_loss=False,  # No loss - this is conditioning
            enable_cfg=True,  
        )

        # Add target image
        target_img = pil_img2rgb(Image.open(io.BytesIO(row["image_list"][-1])))
        data = self._add_image(
            data,
            target_img,
            need_loss=True,  # Compute loss - this is what we want to generate
            need_vae=True,  
            need_vit=False,  
            enable_cfg=False,  
        )

        return data
```

My flags for fine-tuning:
```
  --freeze_vae True \
  --freeze_llm False \
  --freeze_vit False \
  --freeze_und False \
  --visual_gen True \
  --visual_und False \
```

**Comment by Andy1621** [comment]
Hi, here is a example dataset for OmniGen:

```python3
class OmniGenIterableDataset(InterleavedBaseIterableDataset, ParquetStandardIterableDataset):

    def parse_row(self, row):
        image_num = len(row["image"])

        data = self._init_data()
        for idx in range(image_num):
            if idx != image_num - 1:
                data = self._add_image(
                    data, pil_img2rgb(Image.open(io.BytesIO(row["image"][idx]))),
                    need_loss=False, need_vae=True, need_vit=True
                )
            else:
                data = self._add_text(data, row["caption"], need_loss=False)
                data = self._add_image(
                    data, pil_img2rgb(Image.open(io.BytesIO(row["image"][idx]))),
                    need_loss=True, need_vae=False, need_vit=False
                )

        return data
```

**Comment by eternaldolphin** [comment]
Image separator use <"image"> or "<img>|image_n|<img>" or just |image_n|?

**Comment by Andy1621** [comment]
@eternaldolphin Just use the original text of OmniGen dataset

**Comment by Plantian** [comment]
> Hi, I want to congratulate you on your hard work! I saw this[ issue](https://github.com/ByteDance-Seed/Bagel/issues/83) where you explain how to take two images as input, like Omnigen.
> 
> I'm trying to fine-tune Bagel specifically for this kind of task, and so far, I have created a custom dataloader that takes two input images and one input text to generate a target image.
> 
> What I'd like to know is if this is the correct way to fine-tune the model for this type of task, or if I'm missing something or misunderstanding Bagel's architecture. I would greatly appreciate it if you could share the code you used to train the model with multiple input images.
> 
> ```
> class EditingIterableDataset(
>     InterleavedBaseIterableDataset, ParquetStandardIterableDataset
> ):
>     """
>     Custom dataset for editing that ensures a deterministic sequence:
>     Template → Reference → Instruction → Target
> 
>     This guarantees the model always sees both input images before generating the target.
>     """
> 
>     def parse_row(self, row):
>         """
>         Deterministic parsing for food editing:
>         - Always processes Image1 first (conditioning)
>         - Always processes Image2 second (conditioning)
>         - Always processes instruction third (conditioning)
>         - Always generates target image last (with loss)
>         """
>         # Initialize data structure
>         data = self._init_data()
> 
>         # Add Image 1
>         template_img = pil_img2rgb(Image.open(io.BytesIO(row["image_list"][0])))
>         data = self._add_image(
>             data,
>             template_img,
>             need_loss=False,  # No loss
>             need_vae=True, 
>             need_vit=True, 
>             enable_cfg=True,  
>         )
> 
>         # Add Image 2
>         reference_img = pil_img2rgb(Image.open(io.BytesIO(row["image_list"][1])))
>         data = self._add_image(
>             data,
>             reference_img,
>             need_loss=False,  # No loss - this is conditioning input
>             need_vae=True,  
>             need_vit=True,  
>             enable_cfg=True,  
>         )
> 
>         # Add instruction
>         instruction = row["instruction_list"][0]  # Direct access - simple list format
>         
>         data = self._add_text(
>             data,
>             instruction,
>             need_loss=False,  # No loss - this is conditioning
>             enable_cfg=True,  
>         )
> 
>         # Add target image
>         target_img = pil_img2rgb(Image.open(io.BytesIO(row["image_list"][-1])))
>         data = self._add_image(
>             data,
>             target_img,
>             need_loss=True,  # Compute loss - this is what we want to generate
>             need_vae=True,  
>             need_vit=False,  
>             enable_cfg=False,  
>         )
> 
>         return data
> ```
> 
> My flags for fine-tuning:
> 
> ```
>   --freeze_vae True \
>   --freeze_llm False \
>   --freeze_vit False \
>   --freeze_und False \
>   --visual_gen True \
>   --visual_und False \
> ```

hello , i am doing the similar task, are your result clear and not blur?
